harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,440 @@
+        self._metric_collectors = [
+            _GPU(),
+            _CPU(),
+            _Memory(),
+            _Disk(),
+            _Network(),
+        ]
+        self._should_exit = threading.Event()
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        collection_stop_ts = time.time() + self._max_collection_s
+        while not self._should_exit.is_set():
+            if time.time() > collection_stop_ts:
+                # Max collection time reached. Notify queue consumers and exit.
+                self._metrics_queue.put(None)
+                return
+
+            next_collection_ts = time.time() + self._collection_interval
+            self._collect_metrics()
+            wait_ts = max(next_collection_ts - time.time(), 0)
+            self._should_exit.wait(timeout=wait_ts)
+
+    def stop(self) -> None:
+        self._should_exit.set()
+
+    def _collect_metrics(self) -> None:
+        for collector in self._metric_collectors:
+            timestamp = datetime.datetime.now(datetime.timezone.utc)
+            metrics = collector.collect()
+            if not metrics:
+                continue
+            group_metrics = _Metric(group=collector.group, metrics=metrics, timestamp=timestamp)
+            self._metrics_queue.put(group_metrics)
+
+
+class _Shipper(threading.Thread):
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        metrics_queue: queue.Queue,
+        flush_interval: int = 10,
+    ):
+        self._metrics_queue = metrics_queue
+        self._flush_interval = flush_interval
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        self._metrics: typing.List[_Metric] = []
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        while True:
+            metrics: typing.List[_Metric] = []
+            next_flush_ts = time.time() + self._flush_interval
+
+            # Pop messages from queue until the next scheduled flush time.
+            while True:
+                now = time.time()
+                timeout = next_flush_ts - now
+
+                if timeout <= 0:
+                    break
+
+                try:
+                    msg = self._metrics_queue.get(timeout=timeout)
+                    if msg is None:
+                        # Received shutdown message, ship accumulated metrics and exit.
+                        self._ship_metrics(metrics=metrics)
+                        return
+
+                    metrics.append(msg)
+                except queue.Empty:
+                    # Timeout reached, no messages in queue.
+                    break
+
+            self._ship_metrics(metrics=metrics)
+
+    def stop(self) -> None:
+        self._metrics_queue.put(None)
+
+    def _ship_metrics(self, metrics: typing.List[_Metric]) -> None:
```

Comment:
nit: it's unusual in our codebase to dereference from `typing` rather than using the symbols directly. It's like the one exception the style guide exlpicitly calls out.

Reply:
changed it